Add instruction details to cstest logging

hainest opened the pull request:
**Your checklist for this pull request**
- [x] I've documented or updated the documentation of every API function and struct this PR changes.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)

**cstest.py**

Now prints the stringified instruction.

before:
```
INFO  - Run test: Interrupt instructions, 64-bit decode mode -- TestInput { arch: x86, options: ['CS_OPT_DETAIL', 'CS_MODE_64'], addr: 0, bytes: [ 0xcc,0xcd,0x80,0xf1 ] }
WARNING - CS_ARCH: Capstone doesn't have the attribute 'x86'
ERROR - opcode: 241 != 242
INFO  - TestResult.FAILED
```

after:
```
INFO  - Run test: Interrupt instructions, 64-bit decode mode -- TestInput { arch: x86, options: ['CS_OPT_DETAIL', 'CS_MODE_64'], addr: 0, bytes: [ 0xcc,0xcd,0x80,0xf1 ] }
WARNING - CS_ARCH: Capstone doesn't have the attribute 'x86'
ERROR - opcode: 241 != 242
ERROR - Failed instruction: <CsInsn 0x3 [f1]: int1 >
INFO  - TestResult.ERROR
```
---
**cstest.c**

Now prints the test name and the stringified instruction. If a test has no name, it prints the default "(null)". I can change that, if preferred.

before:
```
[ RUN      ] /x86.yaml - TC #11: TestInput { arch: x86, options: [CS_OPT_DETAIL, CS_MODE_64], addr: 0x0, bytes: 0xcc, 0xcd, 0x80, 0xf1 }
actual->opcode[i] != expected->opcode[i]: 241 != 242
[  ERROR   ] --- test_expected_detail(handle, &insns[i], expec_data->details)
[   LINE   ] --- /path/to/capstone/suite/cstest/src/test_case.c:280: error: Failure!
```

after:
```
[ RUN      ] /x86.yaml - TC #11: TestInput { name: Interrupt instructions, 64-bit decode mode, arch: x86, options: [CS_OPT_DETAIL, CS_MODE_64], addr: 0x0, bytes: 0xcc, 0xcd, 0x80, 0xf1 }
actual->opcode[i] != expected->opcode[i]: 241 != 242
[  ERROR   ] --- Failed instruction: int1 (0xf1)
ERROR: test_expected_detail
[   LINE   ] --- /path/to/capstone/suite/cstest/src/test_case.c:313: error: Failure!
```

hainest committed:
Update cstest.py logging

hainest committed:
Update cstest.c logging

Rot127 commented on suite/cstest/src/test_case.c:
```diff
@@ -75,11 +75,12 @@ char *test_input_stringify(const TestInput *test_input, const char *postfix)
+		msg, msg_len,
+		"%sTestInput { name: %s, arch: %s, options: %s, addr: 0x%" PRIx64
+		", bytes: %s }",
+		postfix, test_input->name, test_input->arch, opt_seq,
```
```suggestion
		prefix, test_input->name, test_input->arch, opt_seq,
```

Please rename while you are at it.

hainest replied:
Updated.

Rot127 commented on bindings/python/cstest_py/src/cstest_py/cstest.py:
```diff
@@ -229,42 +229,34 @@ def __init__(self, expected_dict: dict):
-                return TestResult.FAILED
+            def _check(res: bool):
+                if not res:
+                    raise Exception(f"Failed instruction: {a_insn}")
```
Why raising an exception instead of returning `TestResult.FAILED`?
This would stop other tests from execute.

Rot127 replied:
If you want to exit early on the first failing test case, please make it optional via some flag (see `cstest.py::parse_args()`).
And please exit gracefully, not with an exception then.

hainest replied:
> Why raising an exception instead of returning `TestResult.FAILED`?

~I was following the convention of the first check in the function (`if len(actual_insns) != len(self.insns)`).~

I just realized *I* set that convention. Reverted.

> This would stop other tests from execute.

These exceptions are immediately caught in `TestCase.test` where this is called.

---

I just realized that this transformation changes the final status from `FAILED` to `ERROR`. I need to fix that. It will also address one of the changes mentioned in https://github.com/capstone-engine/capstone/pull/2787#discussion_r2367427554.

Rot127 commented on bindings/python/cstest_py/src/cstest_py/cstest.py:
```diff
@@ -293,27 +285,17 @@ def test(self) -> TestResult:
-
-        try:
             return self.expected.compare(insns, self.input.arch_bits)
         except Exception as e:
-            log.error(f"Compare expected failed with: {e}")
-            traceback.print_exc()
-            return TestResult.ERROR
+            return _log_err("Setup", e)
```
Please don't merge those steps.

These are the bindings, so we need to distinguish if the setup, decode, or comparing step failed.

hainest replied:
Good catch! My first transformation preserved that, but I clearly lost track of that important item. I fully reverted this since `TestExpected.compare` now returns a `TestResult` instead of throwing.

Rot127 reviewed (changes requested):
Thanks, definitely a often requested addition!

@Antelox If you have a minute feel free to check the Python changes as well.

hainest committed:
Change exception to log+return in TestExpected.compare

The prior change in TestCase.test converts the exception to
TestResult.FAILED which doesn't preserve the status here.

hainest committed:
Restore calling sequence in TestCase.test

The transformation didn't preserve error reporting. It also is
unnecessary now that TestExpected.compare returns a TestResult again.

hainest committed:
Convert exception to return for length check in TestExpected.compare

The callsite in TestCase.test no longer converts exceptions to a
TestResult.

hainest committed:
Rename local variable 'postfix' to 'prefix' in test_input_stringify

This better reflects what it actually is.

hainest commented:
I reverted the change to the mismatched number of instructions message. It broke the cstest differently than cstest_py that I fixed. I have confirmed both are now producing the same test results:

```
$ ctest --output-on-failure
Test project /path/to/capstone-engine/capstone/build
      Start  1: MCTests
 1/13 Test  #1: MCTests ..........................   Passed    1.69 sec
      Start  2: DetailTests
 2/13 Test  #2: DetailTests ......................   Passed    0.04 sec
      Start  3: IssueTests
 3/13 Test  #3: IssueTests .......................   Passed    0.02 sec
      Start  4: FeaturesTests
 4/13 Test  #4: FeaturesTests ....................   Passed    0.01 sec
      Start  5: unit_cstest
 5/13 Test  #5: unit_cstest ......................   Passed    0.00 sec
      Start  6: integration_cstest
 6/13 Test  #6: integration_cstest ...............   Passed    0.06 sec
      Start  7: legacy_test_skipdata
 7/13 Test  #7: legacy_test_skipdata .............   Passed    0.00 sec
      Start  8: legacy_test_iter
 8/13 Test  #8: legacy_test_iter .................   Passed    0.01 sec
      Start  9: legacy_test_customized_mnem
 9/13 Test  #9: legacy_test_customized_mnem ......   Passed    0.00 sec
      Start 10: integration_compat_headers
10/13 Test #10: integration_compat_headers .......   Passed    0.00 sec
      Start 11: integration_test_litbase
11/13 Test #11: integration_test_litbase .........   Passed    0.00 sec
      Start 12: unit_sstream
12/13 Test #12: unit_sstream .....................   Passed    0.00 sec
      Start 13: unit_utils
13/13 Test #13: unit_utils .......................   Passed    0.00 sec

100% tests passed, 0 tests failed out of 13


$ cstest_py ../tests
...
-----------------------------------------
Test run statistics

Valid files: 1649
Invalid files: 0
Errors: 0

Test cases:
	Total: 82468
	Successful: 81538
	Skipped: 930
	Failed: 0
-----------------------------------------
```

Rot127 reviewed (approved):
Lovely! Thanks a lot!
